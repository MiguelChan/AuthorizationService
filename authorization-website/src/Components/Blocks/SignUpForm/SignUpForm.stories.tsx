import React from 'react';
import {
  Story,
  Meta,
} from '@storybook/react';
import { SignUpForm, SignUpFormProps } from './SignUpForm';

export default {
  title: 'Components/Blocks/SignUpForm',
  component: SignUpForm,
  argTypes: {
    onSignUpRequestCreatedListener: {
      defaultValue: 'onSignUpRequestCreatedListener',
      description: 'onSignUpRequestCreatedListener',
      name: 'onSignUpRequestCreatedListener',
    },
  },
} as Meta;

const Template: Story<SignUpFormProps> = (args) => <SignUpForm {...args} />;

export const Primary = Template.bind({});