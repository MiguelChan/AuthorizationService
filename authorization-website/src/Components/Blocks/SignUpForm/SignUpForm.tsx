import { Button, Grid, TextField, Typography, useTheme } from '@mui/material';
import React from 'react';
import { useFormik } from 'formik';
import * as yup from 'yup';
import 'yup-phone';
import {
  SignUpRequest,
} from '../../../Models';

const validationSchema = yup.object({
  firstName: yup
    .string()
    .required('First name is required'),
  lastName: yup
    .string()
    .required('Last name is required'),
  phoneNumber: yup
    .string()
    .phone()
    .required(),
  email: yup
    .string()
    .email()
    .required('Email is required'),
  password: yup.string()
    .min(8, 'Password must be 8 characters minimum')
    .max(20, 'Password cant be longer than 20 characters')
    .required('Password is required'),
  passwordConfirm: yup
    .string()
    .oneOf([yup.ref('password'), null], 'Passwords must match'),
});

export interface SignUpFormProps {
  onSignUpRequestCreatedListener: (signUpRequest: SignUpRequest) => void;
}

export const SignUpForm: React.FunctionComponent<SignUpFormProps> = ({
  onSignUpRequestCreatedListener,
}) => {

  const currentTheme = useTheme();

  const formik = useFormik({
    initialValues: {
      firstName: '',
      lastName: '',
      phoneNumber: '',
      password: '',
      passwordConfirm: '',
      email: '',
    },
    validationSchema,
    onSubmit: (values) => {
      onSignUpRequestCreatedListener(values);
    },
  });

  const renderButtons = (): React.ReactElement => (
    <>
      <Grid container spacing={2} marginTop={currentTheme.spacing(1)}>
        <Grid item xs='auto'>
          <Button 
            color='primary' 
            variant='contained' 
            type='submit'
            onClick={(): any => formik.submitForm()}
          >
            Create Account
          </Button>
        </Grid>
        <Grid item xs='auto'>
          <Button color='secondary' variant='contained'>Cancel Account Creation</Button>
        </Grid>
      </Grid>
    </>
  );

  const renderTextField = (
    id: string, 
    name: string, 
    label: string,
    currentValue: string,
    hasError: any,
    errorMessage: any,
  ): React.ReactElement => {
    return (
      <TextField 
        fullWidth
        id={id}
        name={name}
        label={label}
        value={currentValue}
        onChange={formik.handleChange}
        error={hasError}
        helperText={errorMessage}
      />
    );
  };

  const renderForm = (): React.ReactElement => (
    <>
      <Grid container spacing={3}>
        <Grid item xs={6}>
          {renderTextField(
            'firstName', 
            'firstName', 
            'First name', 
            formik.values.firstName, 
            formik.touched.firstName && Boolean(formik.errors.firstName), 
            formik.touched.firstName && formik.errors.firstName)}
        </Grid>
        <Grid item xs={6}>
          {renderTextField(
            'lastName', 
            'lastName', 
            'Last name', 
            formik.values.lastName, 
            formik.touched.lastName && Boolean(formik.errors.lastName), 
            formik.touched.lastName && formik.errors.lastName)}
        </Grid>
        <Grid item xs={6}>
          {renderTextField(
              'phoneNumber', 
              'phoneNumber', 
              'Phone Number', 
              formik.values.phoneNumber, 
              formik.touched.phoneNumber && Boolean(formik.errors.phoneNumber), 
              formik.touched.phoneNumber && formik.errors.phoneNumber)}
        </Grid>
        <Grid item xs={6}>
          {renderTextField(
            'email', 
            'email', 
            'Email ', 
            formik.values.email, 
            formik.touched.email && Boolean(formik.errors.email), 
            formik.touched.email && formik.errors.email)}
        </Grid>
        <Grid item xs={6}>
          {renderTextField(
            'password', 
            'password', 
            'Password', 
            formik.values.password, 
            formik.touched.password && Boolean(formik.errors.password), 
            formik.touched.password && formik.errors.password)}
        </Grid>
        <Grid item xs={6}>
          {renderTextField(
            'passwordConfirm', 
            'passwordConfirm', 
            'Confirm password', 
            formik.values.passwordConfirm, 
            formik.touched.passwordConfirm && Boolean(formik.errors.passwordConfirm), 
            formik.touched.passwordConfirm && formik.errors.passwordConfirm)}
        </Grid>
      </Grid>
    </>
  );

  const renderFormTitle = (): React.ReactElement => (
    <Grid container spacing={3}>
      <Grid item xs={12}>
        <Typography variant='h6' component='div' sx={{ flexGrow: 1 }}>Register to the Auth-Hub</Typography>
      </Grid>
    </Grid>
  );

  return (
    <>
      {renderFormTitle()}
      {renderForm()}
      {renderButtons()}
    </>
  );
};