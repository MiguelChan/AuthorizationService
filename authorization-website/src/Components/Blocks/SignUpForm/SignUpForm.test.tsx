import React from 'react';
import {
  render,
  RenderResult,
} from '@testing-library/react';
import { SignUpForm } from './SignUpForm';

describe('SignUpForm', () => {

  const mockOnSignUpRequestListener = jest.fn();

  afterEach(() => {
    mockOnSignUpRequestListener.mockClear;
  });

  it('Should display the Form', () => {
    const renderResult = setupComponent();
    expect(renderResult.container).toMatchSnapshot();
  });

  const setupComponent = (): RenderResult => {
    return render(<SignUpForm onSignUpRequestCreatedListener={mockOnSignUpRequestListener} />);
  };
});