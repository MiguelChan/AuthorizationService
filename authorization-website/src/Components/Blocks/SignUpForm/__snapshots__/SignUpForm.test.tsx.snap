// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm Should display the Form 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
      >
        Register to the Auth-Hub
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="firstName"
          id="firstName-label"
        >
          First name
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="firstName"
            name="firstName"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-173wfxe"
            >
              <span>
                First name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="lastName"
          id="lastName-label"
        >
          Last name
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="lastName"
            name="lastName"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-173wfxe"
            >
              <span>
                Last name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="phoneNumber"
          id="phoneNumber-label"
        >
          Phone Number
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="phoneNumber"
            name="phoneNumber"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-173wfxe"
            >
              <span>
                Phone Number
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="email"
          id="email-label"
        >
          Email 
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="email"
            name="email"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-173wfxe"
            >
              <span>
                Email 
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="password"
            name="password"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-173wfxe"
            >
              <span>
                Password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="passwordConfirm"
          id="passwordConfirm-label"
        >
          Confirm password
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="passwordConfirm"
            name="passwordConfirm"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-173wfxe"
            >
              <span>
                Confirm password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-5sfu98-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto css-7vx2u5-MuiGrid-root"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Create Account
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto css-7vx2u5-MuiGrid-root"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Cancel Account Creation
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
